Algorithm/QCAlgorithm.History.cs

```diff
@@ -624,34 +626,91 @@ public BaseData GetLastKnownPrice(Security security)
         /// This method is used to keep backwards compatibility for those History methods that expect an ArgumentException to be thrown
         /// when the security and the requested data type do not match
         /// </summary>
-        private IEnumerable<T> GetDataTypedHistory<T>(Symbol symbol, IEnumerable<HistoryRequest> requests)
+        /// <remarks>
+        /// This method will check for Python custom data types in order to call the right Slice.Get dynamic method
+        /// </remarks>
+        private IEnumerable<T> GetDataTypedHistory<T>(IEnumerable<HistoryRequest> requests, Symbol symbol)
             where T : IBaseData
         {
+            var type = typeof(T);
+
             if (requests == null || !requests.Any())
             {
                 throw new ArgumentException($"No history data could be fetched. " +
-                    $"This could be due to the specified security not being of the requested type. Symbol: {symbol} Requested Type: {typeof(T).Name}");
+                    $"This could be due to the specified security not being of the requested type. Symbol: {symbol} Requested Type: {type.Name}");
             }
 
             var slices = History(requests);
 
+            // If T is a custom data coming from Python (a class derived from PythonData), T will get here as PythonData
+            // and not the actual custom type. We take care of this especial case by using a dynamic version of GetDataTypedHistory that
+            // receives the Python type, and we get it from the history requests.
+            if (type == typeof(PythonData))
+            {
+                return GetPythonCustomDataTypeHistory(slices, requests.First().DataType, symbol).OfType<T>();
```

Comment:
- `requests.First()` ideally we should get around iterating twice over the generation of the history requests
- `requests.First().DataType` maybe we can pass the requests as a list to `GetPythonCustomDataTypeHistory` so internally it determines what to use? Similar comment as before, avoiding consumers needing to be smart about it